circuit/algorithms/src/poseidon/hash_many.rs

```diff
@@ -164,16 +164,15 @@ impl<E: Environment, const RATE: usize> Poseidon<E, RATE> {
 
     /// Apply the Maximally Distance Separating (MDS) matrix in-place.
     #[inline]
-    fn apply_mds(&self, state: &mut [Field<E>]) {
-        let mut new_state = Vec::with_capacity(state.len());
+    fn apply_mds(&self, state: &mut [Field<E>], new_state: &mut Vec<Field<E>>) {
         for i in 0..state.len() {
             let mut accumulator = Field::zero();
             for (j, element) in state.iter().enumerate() {
                 accumulator += element * &self.mds[i][j];
             }
             new_state.push(accumulator);
         }
-        state.clone_from_slice(&new_state);
+        state.swap_with_slice(new_state);
```

Comment:
Can we guarantee that `new_state` and state` have the same length?

`clone_from_slice` should never panic

Reply:
Got it. To preserve expected behavior of these functions, could you add a `new_state.clear` to the top of `apply_mds`?